File: src/main/java/de/schramm/royalbash/controller/service/core/Player.java
package de.schramm.royalbash.controller.service.core;

import lombok.Builder;
import lombok.Singular;
import lombok.Value;
import lombok.val;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Value
@Builder(toBuilder = true)
public class Player {

    private final String id;
    private final String name;
    private final int hitpoints;
    private final int resources;
    @Singular("handcard")
    private final List<Card> handcards;
    @Singular("deckcard")
    private final List<Card> deckcards;
    @Singular("depositcard")
    private final List<Card> depositcards;
    @Singular("spot")
    private final List<Spot> spots;

    public Player setHitpoints(int hitpoints) {
        return Player.builder()
                .hitpoints(hitpoints)
                .build();
    }

    public Stream<Spot> getSpots() {
        return spots.stream();
    }

    public Stream<Card> getHandcards() {
        return handcards.stream();
    }

    public Stream<Card> getDeckcards() {
        return deckcards.stream();
    }

    public Player updateSpot(Spot oldSpot, Spot newSpot) {

        val spots = this.spots.stream()
                .map(spot -> spot.equals(oldSpot) ? newSpot : spot)
                .collect(Collectors.toList());

        return this.toBuilder()
                .clearSpots()
                .spots(spots)
                .build();
    }

    public Player discardCards(int amountOfCards) {

        Player player = this;
        for(int iterator = 0; iterator < amountOfCards; iterator++) {
            player = player.discardCard();
        }

        return player;
    }

    public Player drawCards(int amountOfCards) {

        Player player = this;
        for(int iterator = 0; iterator < amountOfCards; iterator++) {
            player = player.drawCard();
        }

        return player;
    }

    public Optional<Card> findHandcard(String cardId) {
        return getHandcards()
                .filter(card -> cardId.equals(card.getId()))
                .findFirst();
    }

    Player removeHandcard(Card handcard) {

        val handcards = getHandcards()
                .filter(ownHandcard -> !ownHandcard.equals(handcard))
                .collect(Collectors.toList());

        return findHandcard(handcard)
                .map(card -> this.toBuilder()
                        .depositcard(card)
                        .clearHandcards()
                        .handcards(handcards)
                        .build())
                .orElse(this);
    }

    Player removeCreature(Creature creature) {

        val spots = getSpots()
                .map(spot -> spot.removeCreature(creature))
                .collect(Collectors.toList());

        return findCreature(creature)
                .map(ownedCreature -> this.toBuilder()
                        .depositcard(ownedCreature)
                        .clearSpots()
                        .spots(spots)
                        .build())
                .orElse(this);
    }

    Player updateCreature(Creature oldCreature, Creature newCreature) {

        val spots = getSpots()
                .map(spot -> spot.updateCreature(oldCreature, newCreature))
                .collect(Collectors.toList());

        return this.toBuilder()
                .clearSpots()
                .spots(spots)
                .build();
    }

    Player reduceResources(int cost) {
        return this.toBuilder()
                .resources(resources - cost)
                .build();
    }

    Stream<Card> getDepositcards() { return depositcards.stream(); }

    boolean hasCard(Card card) {
        return getHandcards()
                .anyMatch(handcard -> handcard.equals(card));
    }

    private Player drawCard() {

        if(deckcards.isEmpty()) {
            return this;
        }

        return this.toBuilder()
                .clearDeckcards()
                .deckcards(deckcards.subList(1, deckcards.size()))
                .handcard(deckcards.get(0))
                .build();
    }

    private Player discardCard() {

        if(handcards.isEmpty()) {
            return this;
        }

        return this.toBuilder()
                .clearHandcards()
                .handcards(handcards.subList(1, handcards.size()))
                .build();
    }

    private Optional<Card> findHandcard(Card card) {
        return getHandcards()
                .filter(card::equals)
                .findFirst();
    }

    private Optional<Creature> findCreature(final Creature creature) {
        return getSpots()
                .map(Spot::getCreature)
                .filter(Optional::isPresent)
                .map(Optional::get)
                .filter(creature::equals)
                .findFirst();
    }
}

File: src/main/java/de/schramm/royalbash/controller/service/core/Player.kt
package de.schramm.royalbash.controller.service.core

import java.util.*
import java.util.Collections.emptyList
import java.util.stream.Collectors
import java.util.stream.Stream

data class Player (
        val id: String,
        val name: String = "Player",
        val hitpoints: Int = 0,
        val resources: Int = 0,
        val handcards: List<Card> = emptyList(),
        val deckcards: List<Card> = emptyList(),
        val depositcards: List<Card> = emptyList(),
        val spots: List<Spot> = emptyList()
) {

    fun setHitpoints(hitpoints: Int): Player {
        return copy(hitpoints = hitpoints)
    }

    fun getSpots(): Stream<Spot> {
        return spots.stream()
    }

    fun getHandcards(): Stream<Card> {
        return handcards.stream()
    }

    fun getDeckcards(): Stream<Card> {
        return deckcards.stream()
    }

    fun getDepositcards(): Stream<Card> {
        return depositcards.stream()
    }

    fun updateSpot(oldSpot: Spot, newSpot: Spot): Player {

        val spots = this.spots.stream()
                .map { spot -> if (spot == oldSpot) newSpot else spot }
                .collect(Collectors.toList())

        return copy(spots = spots)
    }

    fun discardCards(amountOfCards: Int): Player {

        var player = this
        for (iterator in 0 until amountOfCards) {
            player = player.discardCard()
        }

        return player
    }

    fun drawCards(amountOfCards: Int): Player {

        var player = this
        for (iterator in 0 until amountOfCards) {
            player = player.drawCard()
        }

        return player
    }

    fun findHandcard(cardId: String): Optional<Card> {
        return getHandcards()
                .filter { card -> cardId == card.id }
                .findFirst()
    }

    internal fun removeHandcard(handcard: Card): Player {

        val handcards = getHandcards()
                .filter { ownHandcard -> ownHandcard != handcard }
                .collect(Collectors.toList())

        return findHandcard(handcard)
                .map { card -> copy(
                        handcards = handcards,
                        depositcards = depositcards.plus(card)
                ) }
                .orElse(this)
    }

    internal fun removeCreature(creature: Creature): Player {

        val spots = getSpots()
                .map { spot -> spot.removeCreature(creature) }
                .collect(Collectors.toList())

        return findCreature(creature)
                .map { ownedCreature -> copy(
                        depositcards = depositcards.plus(ownedCreature),
                        spots = spots
                ) }
                .orElse(this)
    }

    internal fun updateCreature(oldCreature: Creature, newCreature: Creature): Player {

        val spots = getSpots()
                .map { spot -> spot.updateCreature(oldCreature, newCreature) }
                .collect(Collectors.toList())

        return copy(spots = spots)
    }

    internal fun reduceResources(cost: Int): Player {
        return copy(resources = resources - cost)
    }

    internal fun hasCard(card: Card): Boolean {
        return getHandcards()
                .anyMatch { handcard -> handcard == card }
    }

    private fun drawCard(): Player {
        return if (deckcards.isEmpty()) {
            this
        } else copy(
                deckcards = deckcards.subList(1, deckcards.size),
                handcards = handcards.plus(deckcards[0]))

    }

    private fun discardCard(): Player {
        return if (handcards.isEmpty()) {
            this
        } else copy(handcards = handcards.subList(1, handcards.size))
    }

    private fun findHandcard(card: Card): Optional<Card> {
        return getHandcards()
                .filter { card == it }
                .findFirst()
    }

    fun findCreature(creature: Creature): Optional<Creature> {
        return getSpots()
                .map { it.getCreature() }
                .filter { it.isPresent }
                .map { it.get() }
                .filter { creature == it }
                .findFirst()
    }
}
